Switch backend to use Fastify

PurkkaKoodari opened the pull request:
*re-creating pull #84 by @epelip thanks to GitHub*

Refactors ilmomasiina backend to use Fastify as the underlying REST backend instead of Express + Feathers.

This branch builds on top of changes made in #81 so it needs to be merged first.

Resolves #62

## Why?
Feathers enforces clean API design with less code in general and that's probably the reason why Ilmomasiina was build on top of it. However, the application logic has evolved a lot and the amount of logic between the HTTP API and database has increased. All resources does not have all crud CRUD operations or they are diverged between user and admin endpoints. Currently, core features from Feathers have been disabled (i.e. database adapters and socket.io) and the increasing number of hooks makes it more difficult to implement new features, so switching to a less restrictive framework helps to keep this project maintainable.

Fastify is growing ecosystem for REST APIs and one of its best features is JSON Schema support with integrated request validation. Now, as we have separate `ilmomasiina-models`, it is possible to use TypeBox to create single source of schemas providing JSON schema for the API and also typings for TypeScript and input validation rules for both frontend and backend.

## Changes
- Refactor `ilmomasiina-models` to utilize JSON Schema using TypeBox
- Rewrite routing logic from express & Feathers to Fastify
- Improve audit logging to integrate it better with Fastify
- Re-implement admin authentication
- Change signup edit token to be submitted as a header instead of in query params / body
- Rewrite HTTPS enforcement
- Update docs
- Other minor changes to API schemas

epelip committed:
Refactor backend to use fastify instead of feathers

epelip committed:
Refactor admin authentication

epelip committed:
Fix api paths

epelip committed:
Minor schema fixes

epelip committed:
Improve session authentication

epelip committed:
Fix session refreshing

epelip committed:
Improve audit log implementation to work better with fastify

epelip committed:
Resolve & cleanup TODOs

epelip committed:
Re-organize schemas in ilmomasiina-models

epelip committed:
Improve schema naming & doc strings

epelip committed:
Use new schemas in frontend & components

epelip committed:
Re-add edit token verification

epelip committed:
Include edit token header name in ilmomasiina-models

epelip committed:
Fix HTTPS enforcement

epelip committed:
Resolve historyApiFallback for frontend static files

epelip committed:
Add response compression

epelip committed:
Update docs

epelip committed:
Fix remaining type errors

PurkkaKoodari commented on packages/ilmomasiina-backend/src/app.ts:
```diff
@@ -1,74 +1,46 @@
-  if (config.isAzure || config.trustProxy) {
-    app.set('trust proxy', true);
-  }
+  server.register(fastifyCookie, {
+    // secret: 'my-secret', // for cookies signature
+    // parseOptions: {}, // options for parsing cookies
+  });
```
Should we use cookies anywhere? From my knowledge, we previously only stored JWTs in localStorage. Cookies in general are a pain to handle.

epelip replied:
The cookies are only used for authenticated sessions. Previously the JWT tokens were transferred in request bodies alongside the other data. While implementing the JSON schema for the endpoints it felt a bit anti-pattern to mix the object properties and JWTs together. It was also much easier to construct the schema when JWTs are transferred outside the body or query params, i.e. as a header. I first thought that cookies are just simpler to use at the frontend-side, but handling it manually using localStorage is also an option.

epelip replied:
Discarded cookies and reverted back to use Authorization header in 48b89d7.

PurkkaKoodari commented on packages/ilmomasiina-backend/package.json:
```diff
@@ -80,8 +83,6 @@
     "@types/connect-history-api-fallback": "^1.3.5",
```
```suggestion
```

PurkkaKoodari commented on packages/ilmomasiina-backend/package.json:
```diff
@@ -46,22 +46,25 @@
+    "bcrypt": "^5.0.1",
     "better-npm-run": "^0.1.1",
     "compression": "^1.7.4",
     "connect-history-api-fallback": "^1.6.0",
```
```suggestion
```

PurkkaKoodari commented on packages/ilmomasiina-backend/src/authentication/adminPassword.ts:
```diff
@@ -0,0 +1,12 @@
+  static verifyHash(password: string, hash: string): boolean {
+    return bcrypt.compareSync(password, hash);
+  }
+}
```
This should be a default-export class or two exported functions

PurkkaKoodari commented on packages/ilmomasiina-backend/src/authentication/adminSession.ts:
```diff
@@ -0,0 +1,79 @@
+    httpOnly: true,
+    secure: true,
+    sameSite: 'strict',
+    path: '/api/admin',
```
This definitely breaks if we put Ilmomasiina behind a proxied path (e.g. https://kilta.example/ilmo/...)

epelip replied:
Discarded cookies in 48b89d7.

PurkkaKoodari commented on packages/ilmomasiina-backend/src/bin/server.ts:
```diff
@@ -6,15 +6,19 @@ import config from '../config';
-    : `${config.mailUrlBase}${config.pathPrefix}`;
+  server.listen({
+    port,
+  }).then();
```
```suggestion
  server.listen({ port });
```

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/admin/event/createEvent.ts:
```diff
@@ -0,0 +1,54 @@
+  const event = await Event.sequelize!.transaction(async (transaction) => {
+    const created = await Event.create(
+      {
+        ...request.body,
```
In the current code, we explicitly pick allowed attributes on event creation/update. We should at least make sure nothing can be broken from this.

epelip replied:
With Fastify, if a JSON schema is provided for a route, the request body is validated against it. If the request body does not pass the validation, the route function won't be run, so there is no need to validate or parse the data again. All the validation rules have been moved to the JSON schema.

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/admin/event/createEvent.ts:
```diff
@@ -0,0 +1,54 @@
+        questions: request.body.questions ? request.body.questions.map((q, order) => ({
+          ...q, options: (q.options && q.options.length) ? q.options.join(';') : null, order,
+        })) : [],
+        quotas: request.body.quotas ? request.body.quotas.map((q, order) => ({ ...q, order })) : [],
```
```suggestion
        // add order to questions and stringify answer options
        questions: request.body.questions ? request.body.questions.map((q, order) => ({
          ...q,
          order,
          options: (q.options && q.options.length) ? q.options.join(';') : null,
        })) : [],
        // add order to quotas
        quotas: request.body.quotas ? request.body.quotas.map((q, order) => ({ ...q, order })) : [],
```

PurkkaKoodari reviewed (commented):
Partial review, more to come.

epelip committed:
Apply suggestions from code review

Co-authored-by: PurkkaKoodari <8439661+PurkkaKoodari@users.noreply.github.com>

epelip committed:
Revert to use Authorization header instead of a cookie.

epelip committed:
Minor code style improvement: Default export AdminAuthSession

epelip committed:
Code cleanup: comment out session token renewal endpoint.

Separate issue opened for implementing this feature.

PurkkaKoodari commented on packages/ilmomasiina-models/src/enum.ts:
```diff
@@ -0,0 +1,27 @@
+  TextArea = 'textarea',
+  Number = 'number',
+  Select = 'select',
+  Checkbox = 'checkbox',
```
The naming of enum members should be consistent - probably UPPER_SNAKE_CASE

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-models/src/schema/utils.ts:
```diff
@@ -0,0 +1,15 @@
+/** Utility type that converts fields in the imported API to string. */
+export type StringifyApi<T> = {
```
This file is now duplicated from https://github.com/Tietokilta/ilmomasiina/blob/refactor/use-fastify/packages/ilmomasiina-models/src/utils.ts

epelip replied:
Resolved in 8d88854

PurkkaKoodari committed:
Helper commit for git rename detection

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/admin/event/updateEvent.ts:
```diff
@@ -0,0 +1,182 @@
+      // Check for edit conflicts
+
+      // TODO: Is this necessary anymore?
+      // const expectedUpdatedAt = new Date(request.body.updatedAt ?? '');
```
I think we should detect & warn of edit conflicts even when there is no irreversible damage, since we don't support any kind of change merging.

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/admin/event/updateEvent.ts:
```diff
@@ -0,0 +1,182 @@
+          if (question.existing) {
+            await question.existing.update({
+              ...questionAttribs,
+              // TODO: Splitting by semicolon might cause problems - requires better solution
```
This should be turned into a GH issue, TBH

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/admin/slug/checkSlugAvailability.ts:
```diff
@@ -0,0 +1,20 @@
+    attributes: ['id', 'title'],
+  });
+
+  response.status(200);
```
Is `status(200)` necessary for every request?

epelip replied:
It is not necessary, but in my opinion, it improves readability. For people, who are not so familiar with Fastify, being more explicit with the response status codes makes it easier to understand.

PurkkaKoodari commented on packages/ilmomasiina-backend/src/authentication/adminPassword.ts:
```diff
@@ -0,0 +1,12 @@
+import bcrypt from 'bcrypt';
+
+// eslint-disable-next-line import/prefer-default-export
+export class AdminPasswordAuth {
```
This should be a default-export class or two exported functions

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/utils.ts:
```diff
@@ -0,0 +1,34 @@
+  return typeof s === 'string' ? new Date(s) : s;
+}
+
+export function nullsToUndef<T>(val: T | null): T | undefined {
```
```suggestion
/** Returns `undefined` when passed `null`, and the argument otherwise. */
export function nullsToUndef<T>(val: T | null): T | undefined {
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/utils.ts:
```diff
@@ -0,0 +1,34 @@
+
+export function toDateIfDefined(s?: string | null): Date | null | undefined {
+  return typeof s === 'string' ? new Date(s) : s;
+}
```
This doesn't seem to be used anywhere, and it's unclear if it differs from the above function.

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/utils.ts:
```diff
@@ -0,0 +1,34 @@
+
+  // @ts-ignore
+  return o;
+}
```
- Bug: Return `s` instead of `o`
- Use descriptive variable names (very much because of the above)
- Add a doc comment
- Avoid `@ts-ignore`; use `any` when necessary

```suggestion
/** Recursively converts all `Date` objects in the given object to strings. */
export function stringifyDates<T extends object>(obj: T): StringifyApi<T> {
  const result: any = {};
  for (const [key, val] of Object.entries(obj)) {
    if (val instanceof Date) {
      result[key] = val.toISOString();
    } else if (Array.isArray(val)) {
      result[key] = val.map(stringifyDates);
    } else if (typeof val === 'object' && val !== null) {
      result[key] = stringifyDates(val);
    } else {
      result[key] = val;
    }
  }
  return result;
}
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/utils.ts:
```diff
@@ -0,0 +1,34 @@
+T extends string ? Date : Exclude<T, string> {
+  // @ts-ignore
+  return (typeof s === 'string') ? new Date(s) : s;
+}
```
Avoid `@ts-ignore`. (Yes, this kind of type twiddling is generally a pain in the arse)

```suggestion
export function toDate<T>(s: T): Date | Exclude<T, string> {
  return typeof s === 'string' ? new Date(s) : s as Exclude<T, string>;
}
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/user/createUser.ts:
```diff
@@ -0,0 +1,92 @@
+
+  await auditLogger(AuditEvent.CREATE_USER, {
+    extra: res,
+  });
```
Should be inside the transaction.

epelip replied:
Fixed in deb4b18

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/signup/updateSignup.ts:
```diff
@@ -120,16 +129,13 @@ export default async (
       ...nameFields,
       ...emailField,
-      ..._.pick(data, 'namePublic'),
+      namePublic: !!request.body.namePublic,
```
```suggestion
      namePublic: Boolean(request.body.namePublic),
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/signup/updateSignup.ts:
```diff
@@ -50,26 +47,38 @@ export default async (
+        if (!request.body.answers) { return ''; }
+        const ans = request.body.answers.find((a) => a.questionId === question.id);
+        return ans ? ans.answer : '';
+      })();
```
Same as above, but here we can even do the parsing in a nice "pipeline".

```suggestion
      const answer = request.body.answers
        ?.find((a) => a.questionId === question.id)
        ?.answer
        || '';
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/signup/updateSignup.ts:
```diff
@@ -50,26 +47,38 @@ export default async (
       }
-    }
+      return {};
+    })();
```
I don't personally agree with the inline function trick - in my opinion a `let` with the modification nearby is much more readable. (This could also use a well-placed Lodash `pick`, IMO)

```suggestion
    let nameFields = {};
    if (notConfirmedYet && event.nameQuestion) {
      const { firstName, lastName } = request.body;
      if (!firstName) throw new BadRequest('Missing first name');
      if (!lastName) throw new BadRequest('Missing last name');
      nameFields = { firstName, lastName };
    }

    let emailField = {};
    if (notConfirmedYet && event.emailQuestion) {
      const { email } = request.body;
      if (!email) throw new BadRequest('Missing email');
      emailField = { email };
    }
```

epelip replied:
The original intention was to prevent accidents by defining nameFields and emailField as constants. But I agree using `let` makes it much more cleaner.
Suggested changes accepted in 8d88854.

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/signup/getSignupForEdit.ts:
```diff
@@ -0,0 +1,71 @@
+
+  const event = signup.quota!.event!;
+
+  const response = {
+    signup: {
+      ...stringifyDates(signup.get({ plain: true })),
+      status: signup.status,
+      firstName: nullsToUndef(signup.firstName),
+      lastName: nullsToUndef(signup.lastName),
+      email: nullsToUndef(signup.email),
+      confirmed: !!signup.confirmedAt,
+      answers: signup.answers!,
+      quota: signup.quota!,
+    },
+    event: {
+      ...stringifyDates(event.get({ plain: true })),
+      questions: event.questions!.map((question) => ({
+        ...question.get({ plain: true }),
+        // Split answer options into array
+        options: question.options ? question.options.split(';') : null,
+      })),
+    },
+  };
```
What's the purpose of `nullsToUndef` here?

```suggestion
  const response = stringifyDates({
    signup: {
      ...signup.get({ plain: true }),
      status: signup.status,
      firstName: nullsToUndef(signup.firstName),
      lastName: nullsToUndef(signup.lastName),
      email: nullsToUndef(signup.email),
      confirmed: !!signup.confirmedAt,
      answers: signup.answers!,
      quota: signup.quota!,
    },
    event: {
      ...event.get({ plain: true }),
      questions: event.questions!.map((question) => ({
        ...question.get({ plain: true }),
        // Split answer options into array
        options: question.options ? question.options.split(';') : null,
      })),
    },
  });
```

epelip replied:
The purpose of `nullsToUndef`:
In SQL, natural way to present non-existing optional data is to use `NULL`. In JSON, however, null values won't look that great. The more native way of describing an non-existing optional data is to leave such keys undefined (and not present in the serialized JSON).

epelip replied:
For the suggestion of combining the two `stringifyDates`, I would rather not combine them; the penalty from repetition is not that big in this specific scenario, using such powerful recursive functions in larger scope can accidentally touch to somewhere they shouldn't and also runtime performance is slightly better as the calls are better scoped. TS enforces using types matching with the schema.

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/login/adminLogin.ts:
```diff
@@ -0,0 +1,70 @@
+    request: FastifyRequest<{ Body: schema.AdminLoginSchema }>,
+    reply: FastifyReply,
+  ): Promise<AdminSessionSchema | void> => {
+    // Verify existing token
+    const sessionData = session.verifySession(request);
+
+    // Create a new one
+    const accessToken = session.createSession(sessionData);
```
This should check if the user still exists. Perhaps we should also include some mechanism of invalidating old sessions in the future - e.g., not allow renewing tokens older than the latest password change - but that's future stuff.

epelip replied:
The session should definitely be invalidated in the scenarios you mentioned. I think the proper implementation should execute both the session validation and the other possible queries within the same transaction, to be effective? Implementing that requires more refactoring.

PurkkaKoodari replied:
Verifying user data on every request would lose some of the benefits of JWTs. That being said, it's not a bad idea in theory.

Transaction wrapping should not be necessary. The probability of someone hitting a race condition while their user is being deleted is tiny compared to them just sending the request before the deletion. Extraneous transactions can also cause significant performance issues.

PurkkaKoodari replied:
Resolved in ecb9342

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/login/adminLogin.ts:
```diff
@@ -0,0 +1,70 @@
+      attributes: ['id', 'password', 'email'],
+    });
+
+    if (user && AdminPasswordAuth.verifyHash(request.body.password, user.password)) {
```
Allows for enumerating usernames with a timing attack - not critical but technically an issue

epelip replied:
I think good mitigation requires modifications to the database schema. I don't see other good options than storing usernames hashed in another column and comparing the username by first hashing it and then finding the row using the hash. I think it is better to open a new, separate issue for this.

PurkkaKoodari replied:
How this is typically fixed is calling `verifyHash` with garbage data in case the user doesn't exist.

PurkkaKoodari replied:
Resolved in ecb9342

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/index.ts:
```diff
@@ -0,0 +1,381 @@
+
+  const fastify = fastifyInstance.withTypeProvider<TypeBoxTypeProvider>();
+
+  // Setup admin routes (prefixed with '/admin')
```
Since many of these APIs are moving, we should also move corresponding code under `src/routes/admin` where applicable.

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/user/createUser.ts:
```diff
@@ -0,0 +1,92 @@
+      transaction,
+    });
+
+    if (existing) { throw new Conflict('User with given email already exists'); }
```
```suggestion
    if (existing) throw new Conflict('User with given email already exists');
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/signup/editTokens.ts:
```diff
@@ -0,0 +1,49 @@
+): Promise<void> {
+  const headers = request.headers[EDIT_TOKEN_HEADER]; // NOTE: Fastify converts header names into lower case
+  const header = Array.isArray(headers) ? headers[0] : headers;
+  if (verifyToken(request.params.id, header || '')) { return; }
```
```suggestion
  if (verifyToken(request.params.id, header || '')) return;
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-components/src/api.ts:
```diff
@@ -53,6 +53,8 @@ export default async function apiFetch(uri: string, {
     signal,
   });
 
+  if (response.status === 401 && onUnauthenticated) { onUnauthenticated(); }
```
```suggestion
  if (response.status === 401 && onUnauthenticated) onUnauthenticated();
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-components/src/routes/EditSignup/components/EditForm.tsx:
```diff
@@ -17,15 +17,16 @@ import SignupStatus from './SignupStatus';
 const EditForm = () => {
   const [{ event, signup }] = useStateAndDispatch();
-  const isNew = signup!.confirmedAt === null;
+  // @ts-ignore
```
Seems to work with just the nonnull assertion
```suggestion
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-components/src/utils/signupUtils.ts:
```diff
@@ -35,10 +37,10 @@ function getSignupsAsList(event: AnyEventDetails): SignupWithQuota[] {
   title: string;
   size: number | null;
-  signups: SignupWithQuota[];
+  signups: (AnySignupDetails & { quotaId: QuotaID, quotaName: string, confirmed: boolean })[];
```
Extract this to its own type

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-frontend/src/containers/requireAuth.tsx:
```diff
@@ -6,8 +6,9 @@ import { redirectToLogin } from '../modules/auth/actions';
 
 export default function requireAuth<P>(WrappedComponent: ComponentType<P>) {
-  const RequireAuth = (props: P) => {
+  return (props: P) => {
```
The `const` declaration is here to give the function a name - this also shows up in React dev tools

PurkkaKoodari replied:
![image](https://user-images.githubusercontent.com/8439661/199113060-f383f629-a557-4b13-8c33-3b64180fcd8e.png)

epelip replied:
Reverted in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-components/src/api.ts:
```diff
@@ -35,7 +35,7 @@ export class ApiError extends Error {
 export default async function apiFetch(uri: string, {
   method = 'GET', body, headers, accessToken, signal,
-}: FetchOptions = {}) {
+}: FetchOptions = {}, onUnauthenticated?: () => void) {
```
This should be in `FetchOptions` inside the object, not a new parameter, since we're taking an options object already

epelip replied:
It is possible and might look cleaner, but it actually increases the amount of written code as it is then required to write `onUnauthenticated` in almost every time the function is called. In my opinion, it is more convenient this way.

PurkkaKoodari replied:
I'd still prefer a cleaner API - slightly repetitive calling code is a lesser issue in my opinion.

PurkkaKoodari replied:
Improved in 6aec9fb

PurkkaKoodari commented on packages/ilmomasiina-backend/src/enforceHTTPS/index.ts:
```diff
@@ -0,0 +1,48 @@
+ */
+export default function enforceHTTPS(config: { trustProxy: boolean, isAzure: boolean }) {
+  return async (request: FastifyRequest, reply: FastifyReply): Promise<void> => {
+    if (request.protocol === 'https') { return; }
```
```suggestion
    if (request.protocol === 'https') return;
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/event/getEventDetails.ts:
```diff
@@ -0,0 +1,215 @@
+        firstName: nullsToUndef(signup.firstName),
+        lastName: nullsToUndef(signup.lastName),
+        email: nullsToUndef(signup.email),
+        confirmed: !!signup.confirmedAt,
```
```suggestion
        confirmed: Boolean(signup.confirmedAt),
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-frontend/src/modules/auditLog/actions.ts:
```diff
@@ -34,37 +35,37 @@ export type AuditLogActions =
   const queryString = _.entries(query)
-    .filter(([, value]) => value)
-    .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`)
+    .filter(([, value]) => !!value)
```
No need to cast to boolean, `filter` does that for us
```suggestion
    .filter(([, value]) => value)
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-frontend/src/routes/AuditLog/AuditLogActionFilter.tsx:
```diff
@@ -2,27 +2,32 @@ import React from 'react';
 
   const onChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
-    dispatch(setAuditLogQueryField('action', e.target.value));
+    const event = e.target.value;
+    dispatch(setAuditLogQueryField(
+      'action',
+      (Object.values(AuditEvent) as string[]).includes(event) ? event as unknown as [AuditEvent] : undefined,
+    ));
```
Doesn't crash if the cast is erroneous, so keep it simple. (This suggestion also needs the changes to `src/schema/auditLog/index.ts` and `src/routes/auditlog/getAuditLogs.ts`)
```suggestion
    const event = e.target.value as AuditEvent;
    dispatch(setAuditLogQueryField('action', event));
```

PurkkaKoodari replied:
Resolved in 1f8f0ee

PurkkaKoodari commented on packages/ilmomasiina-models/src/schema/auditLog/index.ts:
```diff
@@ -0,0 +1,65 @@
+  ip: Type.Optional(Type.String({
+    title: 'filter events by IP address',
+  })),
+  action: Type.Optional(
+    Type.Array( // TODO: Requires Ajv to be in coerce array mode
+      attributes.auditEventAction,
+      {
+        title: 'filter events by actions',
+      },
+    ),
+  ),
```
Comes from a query string so there's no sense in making it an array. Don't know if we can assign a title this way, though.
```suggestion
  action: Type.Optional(attributes.auditEventAction),
```

epelip replied:
That's true, as we assume to get only one action to filter with. However, parsing query string attributes is a little ambiguous business: what to do when somebody makes a request with `action` defined twice? There is no universal agreement on which one of them to choose. To make the API more user-friendly, both of them should be considered as valid action filters, and that's why `action` is defined here as an array. Naturally, both of the requested action types should be included in the query result.

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/auditlog/getAuditLogs.ts:
```diff
@@ -0,0 +1,71 @@
+  request: FastifyRequest<{ Querystring: schema.AuditLoqQuery }>,
+  response: FastifyReply,
+): Promise<schema.AuditLogResponse> {
+  let where: WhereOptions<AuditLog>[] = [];
```
For whatever reason this had broken for me. Don't know how it worked before.
```suggestion
  let where: WhereOptions<AuditLog> = [];
```

epelip replied:
Should work without modifications if `action` query param is defined to be an array.

PurkkaKoodari replied:
Seems that this is something that's not detected by TypeScript 4.3 (which I pinned in `package.json` due to a previous issue) but is detected by TypeScript 4.8 (which VS Code uses by default). I went through the type definitions last time and I think they indeed didn't match this one.

We should probably fix it to allow upgrading TypeScript later.

PurkkaKoodari replied:
Resolved in 199fa35

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/auditlog/getAuditLogs.ts:
```diff
@@ -0,0 +1,71 @@
+  if (request.query.action) {
+    where = [
+      ...where,
+      { action: { [Op.in]: request.query.action } },
```
(old issue) Since `action` comes from a query string, `in` makes little sense.
```suggestion
      { action: request.query.action },
```

epelip replied:
That's true when `action` is not an array.
See the older issue for better explanation.

PurkkaKoodari commented on packages/ilmomasiina-models/src/attrs/signup.ts:
```diff
@@ -0,0 +1,31 @@
+// Include the same attributes from Event as /api/events.
```
These don't seem to be used anywhere anymore, and should be deleted.

epelip replied:
Removed in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-models/src/attrs/event.ts:
```diff
@@ -0,0 +1,139 @@
+/** User */
```
Some of these are still used, some aren't. Should probably be cleaned up.

epelip replied:
Unused definitions removed in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-models/src/schema/event/index.ts:
```diff
@@ -0,0 +1,112 @@
+  eventID,
+  eventIdentity,
+  eventSlug,
+  userEventAttributesBasic, userEventAttributesExtended,
```
```suggestion
  userEventAttributesBasic,
  userEventAttributesExtended,
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-models/src/schema/event/index.ts:
```diff
@@ -0,0 +1,112 @@
+  questionsUpdate,
+} from '../question';
+import {
+  quotasCreate, quotasUpdate,
```
```suggestion
  quotasCreate,
  quotasUpdate,
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-models/src/schema/eventList/index.ts:
```diff
@@ -0,0 +1,30 @@
+/** Describes request body for sending information about multiple events for admin */
+export const adminEventListSchema = Type.Array(
+  adminEventListItemSchema,
+  {
+
+  },
```
```suggestion
  {},
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari commented on packages/ilmomasiina-models/src/schema/eventList/index.ts:
```diff
@@ -0,0 +1,30 @@
+/** Describes request body for sending information about multiple events for user */
+export const userEventListSchema = Type.Array(
+  userEventListItemSchema,
+  {
+
+  },
```
```suggestion
  {},
```

epelip replied:
Resolved in 8d88854

PurkkaKoodari reviewed (commented):
Very, very good stuff and thanks for the huge effort! In addition to the comments, a few general things:
- Consistency: some code seems to use `http-errors` and some uses `fastify-sensible`
- Consistency: some places `import * as schema` and others `import { Foo, Bar }` directly
- Lots of repetition of `request.xxx.yyy` - some well-placed destructurings would be good here

epelip committed:
Assess review feedback: cleanup code

epelip committed:
Assess review feedback: ensure atomicity while creating a new user

PurkkaKoodari commented on packages/ilmomasiina-backend/src/routes/login/adminLogin.ts:
```diff
@@ -0,0 +1,70 @@
+}
+
+export function addSessionValidationHook(session: AdminAuthSession, fastify: FastifyInstance): void {
+  fastify.decorateRequest('userID', null);
```
What's the purpose of this line?

PurkkaKoodari committed:
General cleanup and fixes to Fastify conversion

- Make route code paths match API paths
- Clarify API route setup by prefix option next to .register call
- Remove unnecessary eslint comment
- Fix WhereOptions typing for newer TypeScript compat
- Re-enable edit conflict detection
- Move @types to devDependencies in backend

PurkkaKoodari committed:
Move more audit logging inside transactions

Sequelize requires the transaction object to be explicitly passed to all
queries.

PurkkaKoodari committed:
Remove redundant stringifyDates

PurkkaKoodari committed:
Re-add configurable CORS (5573a7c)

PurkkaKoodari commented:
FYI: I've rebased this on top of #90 in [`wip/fastify-rebase`](https://github.com/Tietokilta/ilmomasiina/tree/wip/fastify-rebase) and am adding some changes from that PR there.

PurkkaKoodari committed:
Move most ilmomasiina-models exports to index.ts

PurkkaKoodari committed:
Improve security of authentication
- Check that user still exists when renewing session
- Mitigate user enumeration: call createHash() when user doesn't exist

PurkkaKoodari committed:
Refactor 401 handling into own function

PurkkaKoodari committed:
Simplify type cast in AuditLogActionFilter

PurkkaKoodari committed:
Typing fixes and documentation for signupUtils

PurkkaKoodari committed:
Clean up typebox schemas

- Renamed schemas/types consistently
- Used title and description as recommended in JSON Schema
- Added comments to exported schemas/types
- Refactored types to reduce overall clutter

PurkkaKoodari committed:
Fix fastify route initialization

PurkkaKoodari committed:
Eliminate import cycles

PurkkaKoodari committed:
Update TypeScript and ESLint

PurkkaKoodari committed:
Add config options to tsconfigPathsResolverPlugin

PurkkaKoodari committed:
Fix project references in ESLint

epelip reviewed (approved):
LGTM :+1:

epelip committed:
Respond with a deflated response body by default
(i.e. when a client requests unsupported encoding)
